src/libinjector/debug_helpers.c

```diff
@@ -0,0 +1,189 @@
+ * By sending these changes, it is understood (unless you specify          *
+ * otherwise) that you are offering unlimited, non-exclusive right to      *
+ * reuse, modify, and relicense the code.  DRAKVUF will always be          *
+ * available Open Source, but this is important because the inability to   *
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ * This file was created by Manorit Chawdhry.                              *
+ * It is distributed as part of DRAKVUF under the same license             *
+ ***************************************************************************/
+
+
+#include <libinjector/debug_helpers.h>
+
+void print_hex(const char* array, size_t len)
+{
+    PRINT_DEBUG("Total length: %ld\n", len);
+    PRINT_DEBUG("Data: \n");
+    for (size_t i=0; i<len; i++)
+    {
+        PRINT_DEBUG("%02x ", *(array + i) & 0xff);
+    }
+    PRINT_DEBUG("\n");
+}
+
+void print_stack(drakvuf_t drakvuf, drakvuf_trap_info_t* info, addr_t addr)
+{
+    PRINT_DEBUG("Stack\n");
+    vmi_instance_t vmi = drakvuf_lock_and_get_vmi(drakvuf);
+
+    for (int i=0; i < 16; i++)
+    {
+        ACCESS_CONTEXT(ctx,
+            .translate_mechanism = VMI_TM_PROCESS_PID,
+            .pid = info->proc_data.pid,
+            .addr = (addr + i*8)
+        );
+        addr_t val = 0;
+        vmi_read_64(vmi, &ctx, &val);
+        if ((i%4)==0)
+            PRINT_DEBUG("\n%016lx:", addr + (i/4)*32);
+        PRINT_DEBUG(" %016lx", val);
+    }
+    PRINT_DEBUG("\n");
+
+    drakvuf_release_vmi(drakvuf);
+}
+
+static char* repeat_str (const char* str, size_t count)
+{
+    if (count == 0) return NULL;
+    char* ret = malloc (strlen (str) * count + count);
+    if (ret == NULL) return NULL;
+    strcpy (ret, str);
+    while (--count > 0)
+    {
+        strcat (ret, " ");
+        strcat (ret, str);
+    }
+    return ret;
+}
+
+void print_registers(drakvuf_trap_info_t* info)
+{
+    const char* fmt_base = "%s:\t%016lx\n";
+    char* fmt= repeat_str(fmt_base, 24);
+    PRINT_DEBUG(fmt,
+        "rax",    info->regs->rax,
```

Comment:
I'd prefer to do it with:

```
PRINT_DEBUG(fmt, "rax",    info->regs->rax);
PRINT_DEBUG(fmt, "rcx",    info->regs->rcx);
...
```

This way we can remove `repeat_str`.

Reply:
Done